text/0000-inline-asm.md

```diff
@@ -0,0 +1,869 @@
+
+Let's take our previous example of an `add` instruction:
+
+```rust
+let mut a = 4;
+let b = 4;
+unsafe {
+    asm!(
+        "add {0}, {1}",
+        inlateout(reg) a, in(reg) b,
+        flags(pure, nomem, nostack)
+    );
+}
+assert_eq!(a, 8);
+```
+
+Flags can be provided as an optional final argument to the `asm!` macro. We specified three flags here:
+- `pure` means that the asm code has no observable side effects and that its output depends only on its inputs. This allows the compiler optimizer to call the inline asm fewer times or even eliminate it entirely.
+- `nomem` means that the asm code does not read or write to memory. By default the compiler will assume that inline assembly can read or write any memory address that is accessible to it (e.g. through a pointer passed as an operand, or a global).
+- `nostack` means that the asm code does not push any data onto the stack. This allows the compiler to use optimizations such as the stack red zone on x86_64 to avoid stack pointer adjustments.
+
+These allow the compiler to better optimize code using `asm!`, for example by eliminating pure `asm!` blocks whose outputs are not needed.
+
+See the reference for the full list of available flags and their effects.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Inline assembler is implemented as an unsafe macro `asm!()`.
+The first argument to this macro is a template string literal used to build the final assembly.
+The following arguments specify input and output operands.
+When required, flags are specified as the final argument.
+
+The following ABNF specifies the general syntax:
+
+```
+dir_spec := "in" / "out" / "lateout" / "inout" / "inlateout"
+reg_spec := <arch specific register class> / "<arch specific register name>"
+operand_expr := expr / "_" / expr "=>" expr / expr "=>" "_"
+reg_operand := dir_spec "(" reg_spec ")" operand_expr
+operand := reg_operand / "imm" const_expr / "sym" path
+flag := "pure" / "nomem" / "readonly" / "preserves_flags" / "noreturn"
+flags := "flags(" flag *["," flag] ")"
+asm := "asm!(" format_string *("," [ident "="] operand) ["," flags] ")"
+```
+
+[format-syntax]: https://doc.rust-lang.org/std/fmt/#syntax
+
+## Template string
+
+The assembler template uses the same syntax as [format strings][format-syntax] (i.e. placeholders are specified by curly braces). The corresponding arguments are accessed in order, by index, or by name. However, implicit named arguments (introduced by [RFC #2795][rfc-2795]) are not supported.
+
+The assembly code syntax used is that of the GNU assembler (GAS). The only exception is on x86 where the Intel syntax is used instead of GCC's AT&T syntax.
```

Comment:
> The only exception is on x86 where the Intel syntax is used instead of GCC's AT&T syntax.

Can we see an example of Intel syntax being used with this macro?

Reply:
All examples in this RFC use Intel syntax.

Reply:
In order to aid in mechanical translation of code, could we expose the appropriate syntax flags here, so that people *can* ask for AT&T syntax on x86, for instance? (That would make it easy for others to implement an `asm_att!` macro, for instance.)

And, of course, the alternatives section should mention the possibility of using AT&T syntax on all platforms and providing a flag for Intel syntax, in which case people could easily implement an `asm_intel!` macro for Intel syntax.

Reply:
This is tricky to support reliably. In particular, GCC requires that all asm code in a compilation unit use the same syntax, so this would at least exclude the possibility of inline asm support with a GCC backend.

I would prefer if we simply chose a single asm syntax and stuck with it. Also note that this only affects x86 which has 2 asm syntaxes. Every other architecture only has a single standardized asm syntax.

Reply:
Could it be a crate level setting, since each crate is a single translation unit?

Reply:
I don't think that a crate-level setting will be enough if you take LTO into account. And even then, a crate-level setting won't work with inline functions.

Reply:
No need for that.  You can switch syntaxes on-the-fly using assembler directives, `.att_syntax` and `.intel_syntax noprefix`: [example](https://gcc.godbolt.org/z/d28KDo).

Reply:
The issue is more with the register placeholders, since GCC needs to know whether to emit `eax` or `%eax`.

Reply:
Oh... good point.  For the record, this only affects a hypothetical GCC backend, not LLVM, which supports `inteldialect` per asm block.

I suppose such a backend could always tell GCC to compile in Intel mode, and then just add the `%`s and `$`s itself.

Reply:
@Amanieu GCC supports prefixed registers even when in Intel syntax mode, and even with `noprefix`; `noprefix` just makes the prefixes optional. So you can (and should) always emit prefixes like `%` and `$`, always leave GCC in AT&T mode, and just wrap assembly blocks that use Intel syntax in `.intel_syntax noprefix` and `.att_syntax`.